moveit_experimental/jog_arm/include/jog_arm/jog_arm_server.h

```diff
@@ -0,0 +1,228 @@
+#include <moveit/planning_scene/planning_scene.h>
+#include <moveit/planning_scene_interface/planning_scene_interface.h>
+#include <moveit/robot_model_loader/robot_model_loader.h>
+#include <moveit/robot_state/robot_state.h>
+#include <pthread.h>
+#include <ros/ros.h>
+#include <sensor_msgs/JointState.h>
+#include <sensor_msgs/Joy.h>
+#include <std_msgs/Bool.h>
+#include <string>
+#include <tf/transform_listener.h>
+#include <trajectory_msgs/JointTrajectory.h>
+
+namespace jog_arm
+{
+// For jogging calc thread
+void* joggingPipeline(void* threadid);
+
+// For collision checking thread
+void* collisionCheck(void* threadid);
+
+// Shared variables
+geometry_msgs::TwistStamped g_command_deltas;
+pthread_mutex_t g_command_deltas_mutex;
+
+sensor_msgs::JointState g_joints;
+pthread_mutex_t g_joints_mutex;
+
+trajectory_msgs::JointTrajectory g_new_traj;
+pthread_mutex_t g_new_traj_mutex;
+
+bool g_imminent_collision(false);
+pthread_mutex_t g_imminent_collision_mutex;
+
+bool g_zero_trajectory_flag(false);
+pthread_mutex_t g_zero_trajectory_flagmutex;
+
+// ROS subscriber callbacks
+void deltaCmdCB(const geometry_msgs::TwistStampedConstPtr& msg);
+void jointsCB(const sensor_msgs::JointStateConstPtr& msg);
+
+int readParameters(ros::NodeHandle& n);
+
+// ROS params to be read
+struct jog_arm_parameters
+{
+  std::string move_group_name, joint_topic, command_in_topic, command_frame, command_out_topic, planning_frame,
+      warning_topic;
+  double linear_scale, rotational_scale, singularity_threshold, hard_stop_singularity_threshold, low_pass_filter_coeff,
+      publish_period, incoming_command_timeout;
+  bool gazebo, collision_check;
+} g_parameters;
+
+/**
+ * Class LowPassFilter - Filter the joint velocities to avoid jerky motion.
+ */
+class LowPassFilter
+{
+public:
+  LowPassFilter(double low_pass_filter_coeff);
+  double filter(const double& new_msrmt);
+  void reset(double data);
+  double filter_coeff_ = 10.;
+
+private:
+  double prev_msrmts_[3] = { 0., 0., 0. };
+  double prev_filtered_msrmts_[2] = { 0., 0. };
+};
+
+LowPassFilter::LowPassFilter(double low_pass_filter_coeff)
+{
+  filter_coeff_ = low_pass_filter_coeff;
+}
+
+void LowPassFilter::reset(double data)
+{
+  prev_msrmts_[0] = data;
+  prev_msrmts_[1] = data;
+  prev_msrmts_[2] = data;
+
+  prev_filtered_msrmts_[0] = data;
+  prev_filtered_msrmts_[1] = data;
+}
+
+double LowPassFilter::filter(const double& new_msrmt)
+{
+  // Push in the new measurement
+  prev_msrmts_[2] = prev_msrmts_[1];
+  prev_msrmts_[1] = prev_msrmts_[0];
+  prev_msrmts_[0] = new_msrmt;
+
+  double new_filtered_msrmt = (1 / (1 + filter_coeff_ * filter_coeff_ + 1.414 * filter_coeff_)) *
+                              (prev_msrmts_[2] + 2 * prev_msrmts_[1] + prev_msrmts_[0] -
+                               (filter_coeff_ * filter_coeff_ - 1.414 * filter_coeff_ + 1) * prev_filtered_msrmts_[1] -
+                               (-2 * filter_coeff_ * filter_coeff_ + 2) * prev_filtered_msrmts_[0]);
```

Comment:
divide by zero appears to be possible here, a check should be added....

Reply:
OK, I'll ensure that filter_coeff_ is positive.